jgrapht-core/src/main/java/org/jgrapht/alg/tour/NearestInsertionHeuristicTSP.java

```diff
@@ -0,0 +1,326 @@
+ * list of cities and the distances between each pair of cities, what is the
+ * shortest possible route that visits each city exactly once and returns to the
+ * origin city?".
+ * </p>
+ *
+ * <p>
+ * Insertion heuristics are quite straighforward, and there are many variants to 
+ * choose from. The basics of insertion heuristics is to start with a tour of a 
+ * subset of all cities, and then inserting the rest by some heuristic. The 
+ * initial subtour is often a triangle or the convex hull. One can also start 
+ * with a single edge as subtour. This implementation uses the shortest edge by
+ * default as the initial sub-tour.
+ * </p>
+ *
+ * <p>
+ * The implementation of this class is based on: <br>
+ * Nilsson, Christian. "Heuristics for the traveling salesman problem." Linkoping University 38 (2003)
+ * </p>
+ * 
+ * <p>
+ * This implementation can also be used in order to try to improve an existing tour. See method
+ * {@link #improveTour(GraphPath)}.
+ * </p>
+ *
+ * <p>
+ * The runtime complexity of this class is $O(V^2)$.
+ * </p>
+ * 
+ * <p>
+ * This algorithm requires that the graph is complete.
+ * </p>
+*
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Peter Harman
+ */
+public class NearestInsertionHeuristicTSP<V, E> extends
+        HamiltonianCycleAlgorithmBase<V, E>
+        implements HamiltonianCycleImprovementAlgorithm<V, E> {
+
+    private GraphPath<V, E> subtour;
+
+    /**
+     * Constructor. By default a sub-tour is chosen based on the shortest edge
+     */
+    public NearestInsertionHeuristicTSP() {
+        this(null);
+    }
+
+    /**
+     * Constructor
+     *
+     * @param subtour Initial sub-tour, or null to start with shortest edge
+     */
+    public NearestInsertionHeuristicTSP(GraphPath<V, E> subtour) {
+        this.subtour = subtour;
+    }
+
+    /**
+     * Computes a tour using the nearest insertion heuristic.
+     *
+     * @param graph the input graph
+     * @return a tour
+     * @throws IllegalArgumentException if the graph is not undirected
+     * @throws IllegalArgumentException if the graph is not complete
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        if (graph.vertexSet().size() == 1) {
+            return getSingletonTour(graph);
+        }
+        return listToTour(augment(subtour(graph), graph), graph);
+    }
+
+    /**
+     * Improves an existing tour. Using this algorithm will not improve the cost, 
+     * but instead adds any missing vertices to the tour. It also removes
+     * vertices from the tour if not in the graph, so this can be used to update
+     * a tour following changes to a graph.
+     *
+     * @param tour The existing tour
+     * @return A valid and complete tour for the same graph
+     */
+    @Override
+    public GraphPath<V, E> improveTour(GraphPath<V, E> tour) {
```

Comment:
To me, this method is supposed to get a complete tour and improve it using some heuristics. This implementation seems to have no effect in the case we provide a complete tour. 

Suppose we have an instance of the `HamiltonianCycleImprovementAlgorithm`. If we provide a subtour to the `TwoOptHeuristicTSP`, then we'll get the tour of the same size with a possibly reduced length. If we provide a subtour to the `NearestInsertionHeuristicTSP`, then we'll get a tour of different length. So, the behaviour is not the same. 

In general, the functionality of filtering the removed vertices is too application-specific and it might make sense to decouple it from this class.

Reply:
I have documented that the behaviour isn't the same, but I'm happy for it to be changed if it is felt it adds confusion using the same interface for 2 different intents.

Reply:
I think it's better to change it because it would be too confusing otherwise. Technically, there's no improvement done with respect to the tour. We correct the tour to be valid with respect to modifications and incompleteness. At the same, the responsibility of the nearest insertion heuristics is to produce a valid tour given a valid graph using concrete logic.